standard/expressions.md

```diff
@@ -1824,21 +1824,39 @@ A *null_conditional_projection_initializer* is a restriction of *null_conditiona
 
 ### 12.8.9 Null-forgiving expressions
 
-This operator sets the null state ([§8.9.5](types.md#895-nullabilities-and-null-states)) of the operand to “not null”.
+A null-forgiving expression’s value, type, classification ([§12.2](expressions.md#122-expression-classifications))
+and safe-context ([§16.4.12](structs.md#16412-safe-context-constraint)) is the value, type, classification and safe-context of its *primary_expression*.
 
 ```ANTLR
 null_forgiving_expression
-    : primary_no_array_creation_expression suppression
+    : primary_expression null_forgiving_operator
     ;
 
-suppression
+null_forgiving_operator
     : '!'
     ;
 ```
 
-This operator has no runtime effect; it evaluates to the result of its operand, and that result retains that operand’s classification.
+*Note*: The postfix null-forgiving and prefix logical negation operators ([§12.9.4](expressions.md#1294-logical-negation-operator)), while represented by the same lexical token (`!`), are distinct. Only the latter may be overriden ([§15.10](classes.md#1510-operators)), the definition of the null-forgiving operator is fixed. *end note*
+
+<!-- [Placeholder] 12.8.9 A -->
+<!-- [Placeholder] 12.8.9 A End -->
+**The remainder of this subclause, including all of its subclauses, is conditionally normative.**
+
+A compiler which performs static null state analysis ([§8.9.5](types.md#895-nullabilities-and-null-states)) must conform to the following specification.
+
+The null-forgiving operator is a compile time pseudo-operation that is used to inform a compiler’s static null state analysis. It has two uses: to override a compiler’s determination that an expression *maybe null*; and to override a compiler issuing a warning related to nullability.
+
+Applying the null-forgiving operator to an expression for which a compiler’s static null state analysis
+does not produce any warnings is not an error.
+
+#### 12.8.9.1 Overriding a *maybe null* determination
 
-The null-forgiving operator is used to declare that an expression not known to be a value type is not null.
+Under some circumstances a compiler’s static null state analysis may determine that an expression
+has the null state *maybe null* and issue a warning when other information indicates that the
+expression cannot be null. Applying the null-forgiving operator to such an expression informs the
+compiler’s static null state analysis that the null state is in *not null*; which both prevents the warning
+message and informs any ongoing analysis.
```

Comment:
I can't think of better wording, but I can think of weasel words:

```suggestion
message and may inform any ongoing analysis.
```

Reply:
@BillWagner – I think that is better wording and I’ll do it unless others disagree.